public/pages/workflow_detail/workspace_component/utils.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Connection, ReactFlowInstance } from 'reactflow';
+import { IComponentInput } from '../../../../common';
+
+/**
+ * Collection of utility functions for the workspace component
+ */
+
+// Uses DOM elements to calculate where the handle should be placed
+// vertically on the ReactFlow component. offsetTop is the offset relative to the
+// parent element, and clientHeight is the element height including padding.
+// We can combine them to get the exact amount, in pixels.
+export function calculateHandlePosition(ref: any): number {
+  if (ref.current && ref.current.offsetTop && ref.current.clientHeight) {
+    return ref.current.offsetTop + ref.current.clientHeight / 2;
+  } else {
+    return 0;
+  }
+}
+
+// Validates that connections can only be made when the source and target classes align, and
+// that multiple connections to the same target handle are not allowed unless the input configuration
+// for that particular component allows for it.
+export function isValidConnection(
+  connection: Connection,
+  rfInstance: ReactFlowInstance
+): boolean {
+  const sourceHandle = connection.sourceHandle;
+  const targetHandle = connection.targetHandle;
+  const targetNodeId = connection.target;
+  const inputClass = sourceHandle || '';
+  // We store the output classes in a pipe-delimited string. Converting back to a list.
+  const outputClasses = targetHandle?.split('|') || [];
```

Comment:
Did you test here if we have a list of 3 output classes, and a list of 3 target classes, just one of them matching is enough?

Reply:
Yes. Also, I was splitting the target handle instead of the source handle. Updated in latest commit.

Reply:
I got rid of input/output language since that added some confusion.